src/mono/mono/utils/mono-wasm-pagemgr.c

```diff
@@ -0,0 +1,401 @@
+	transition_page_states (MWPM_ALLOCATED_TO_FREE, first_page, page_count);
+}
+
+static inline const char *
+get_state_name (uint8_t state) {
+	switch (state & MWPM_STATE_MASK) {
+		case MWPM_EXTERNAL:
+			return "external";
+		case MWPM_FREE_DIRTY:
+			return "dirty";
+		case MWPM_FREE_ZEROED:
+			return "zeroed";
+		case MWPM_ALLOCATED:
+			return "in use";
+		default:
+			g_assert_not_reached ();
+	}
+}
+
+static uint32_t
+find_n_free_pages_in_range (uint32_t start_scan_where, uint32_t end_scan_where, uint32_t page_count) {
+	if (page_count == 0)
+		return UINT32_MAX;
+
+	uint32_t i = start_scan_where;
+
+	while (i <= end_scan_where) {
+		uint8_t found_obstruction = 0;
+		uint32_t j = i + page_count - 1;
+		if (j > last_controlled_page_index)
+			break;
+
+		// TODO: If we find a free page with a skip count in it, that would indicate
+		//  that there are N sequential free pages left we can claim without doing
+		//  the scan below.
+
+		// Scan backwards from the last candidate page to look for any non-free pages
+		//  the first non-free page we find is the next place we will search from.
+		for (; j >= i; j--) {
+			mwpm_page_state page_state = page_table[j];
+
+			if (!is_page_free (page_state)) {
+				// Skip multiple pages
+				uint32_t skip_count = get_page_skip_count (page_state);
+				i = j + skip_count;
+#ifdef MWPM_LOGGING
+				if (skip_count > 1)
+					g_print (
+						"scan skipping %u %s page(s); new page is #%u with state %s\n",
+						skip_count, get_state_name (page_state),
+						i, get_state_name (page_table[i])
+					);
+#endif
+				found_obstruction = 1;
+				break;
+			}
+		}
+
+		if (found_obstruction)
+			continue;
+
+		// We scanned page_count pages starting from i and they were all free.
+		return i;
+	}
+
+	return UINT32_MAX;
+}
+
+// Scans all controlled pages to look for at least page_count free pages.
+static uint32_t
+find_n_free_pages (uint32_t page_count) {
+	// Start scanning from the beginning. This ensures we will try to grab small allocations
+	//  from the front of the page table, and large allocations from anywhere we can find.
+	// This does make scans slower, but other approaches I tried have much worse fragmentation.
+	uint32_t result = find_n_free_pages_in_range (first_controlled_page_index, last_controlled_page_index, page_count);
+	return result;
+}
+
+static void
+mwpm_init () {
+	mono_os_mutex_init_recursive (&mutex);
+	// Set the entire page table to 'unknown state'. As we acquire pages from sbrk, we will
+	//  set those respective ranges in the table to a known state.
+	memset (page_table, MWPM_EXTERNAL, sizeof(page_table));
+	void *first_controlled_page_address = acquire_new_pages_initialized (MWPM_MINIMUM_PAGE_COUNT);
+	g_assert (first_controlled_page_address);
+	first_controlled_page_index = first_page_from_address (first_controlled_page_address);
+}
+
+static inline void
+mwpm_ensure_initialized () {
+	call_once (&is_initialized, mwpm_init);
```

Comment:
I kind of want us to assert here that we're using coop suspend.
With hybrid suspend I think we could get into a deadlock here if one threads get here and is preemptively suspended by the suspend initiator while inside `mwpm_init`, and then a second thread gets here in cooperative mode and ends up blocking waiting for the first thread to exit the `call_once`.

We would need to enter GC Safe mode around the call to `call_once`.  But that's expensive/not needed on coop GC.

And in theory we should always be using coop GC on wasm.

```c
	call_once (&is_initialized, mwpm_init);
```